### PR TITLE
Enable `useAuthenticator` usage outside <Authenticator />

### PR DESCRIPTION
**Pull request by wlee221**
*Issue #, if available:* #637, #254, #664, #640.

*Description of changes:* This PR enables `useAuthenticator` hook to be used outside `Authenticator`. Also adds `/useAuthenticator` example and e2e tests that'll supplement docs PR #1075.

## Use Case

Main use case is custom routing (#1058), which is not possible today because Authenticator assumes SPA usage and requires Authenticated content to be nested inside `Authenticator`. #1130 puts it well:

> The [Quick Start instructions](https://ui.docs.amplify.aws/components/authenticator?platform=vue#quick-start) are great... if you want the entire app locked down. Beyond that... it's not immediately clear to me (and likely others, too) how I could exclude certain pages from Authenticator. (eg, Home, About, etc.).

## Problem

Internally, we used `useAuthenticator` and `Authenticator.Provider` to share `auth` contexts. To use `useAuthenticator` outside `Authenticator`, we falsely went around this issue by wrapping `<Authenticator.Provider />` around their app:

```tsx
(1) <Authenticator.Provider> 
(2)   <App>
(3)     <Authenticator> 
(4)       <Authenticator.Provider>
(5)         <Authenticator.SignIn />
(6)        ...
```

Problem here was that two providers create two different xstate machines. `useAuthenticator` from `<App />` (2) returned false data because it is referencing the wrong provider (1) instead of (4).

### Wasted re-renders

Another React nuance we overlooked is that `useAuthenticator` changes on every xstate transitions, including form input events! 

This is not only computationally expensive but also can trigger repeated mounting of `<Authenticator />` because `MyApp` (1) encompasses it. This opens up crucial bugs like form losing its values on re-render. [xstate docs](https://xstate.js.org/docs/recipes/react.html#global-state-react-context) explains better than I can:

> React context can be a tricky tool to work with - if you pass in values which change too often, it can result in re-renders all the way down the tree. 

### Sync Provider Context Value 

Now `Authenticator.Provider` detects whether there's another `Provider` above it. If so, it'll just pass that parent value. This resolves multiple xstate machine instantiations.

### Serve static value from `Authenticator.Provider`

Now `Authenticator.Provider` only maintains a static reference to `authMachine` service to prevent re-renders, as suggested from [xstate docs](https://xstate.js.org/docs/recipes/react.html#global-state-react-context). `useAuthenticator` instead listens to state changes and provides respective auth context.

### Split `Slot` logic to internal hook

We've been using `useAuthenticator` to share custom components for Authenticator. This is only needed internally, is static, and is not needed externally. This has been split to `useCustomComponents` hook to focus `Authenticator.Provider` on just `authMachine` contexts. Splitting up complex context value is encouraged to reduce wasted re-renders: https://github.com/facebook/react/issues/15156.

### Add optimization mechanism to useAuthenticator

Even with static served `Authenticator.Provider`, `useAuthenticator` still changes frequently on every state change. This PR adds `useAuthenticatorRoute` and `useAuthenticatorUser` that only triggers re-render whenever `route` or `user` change respectively. This provides developers control over when they want auth updates.

Under the hood, this is done by `selector` optional parameter to base `useAuthenticator`. `selector` is a function that accepts current facade values and returns desired value(s) that should trigger re-render. 

This uses `xstate`'s [`useSelector`](https://xstate.js.org/docs/packages/xstate-react/#useselector-actor-selector-compare-getsnapshot), which provides just what we need: 

>  This hook will only cause a rerender if the selected value changes, ...

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/react/src/components/Authenticator/Authenticator.tsx
Comment (wlee221) [inline packages/react/src/components/Authenticator/Authenticator.tsx:49]: This is a follow up from #1162. We're setting `INIT` event after we get the necessary `machineProps` in `Authenticator`.

File: packages/react/src/components/Authenticator/ConfirmSignUp/ConfirmSignUp.tsx
Comment (wlee221) [inline packages/react/src/components/Authenticator/ConfirmSignUp/ConfirmSignUp.tsx:21]: `useAuthenticator` now returns only contexts related to `auth`!

File: packages/react/src/components/Authenticator/Router/index.tsx
Comment (ErikCH) [inline packages/react/src/components/Authenticator/Router/index.tsx:16]: Glad you made this optional. Always bothered me you had to add the children using `authenticator`

File: packages/e2e/features/ui/components/authenticator/useAuthenticator.feature
Comment (ErikCH) [inline packages/e2e/features/ui/components/authenticator/useAuthenticator.feature:13]: I can try to make sure Vue works in the future here, with this new useAuthenticator feature.

File: packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx
Comment (ErikCH) [inline packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx:96]: Why isn't the selector passed in from the `useAuthenticator` used here? Like this

`const state = useSelector(service, selector, comparator)`

File: packages/e2e/features/ui/components/authenticator/useAuthenticator.feature
Comment (wlee221) [reply packages/e2e/features/ui/components/authenticator/useAuthenticator.feature:13]: Yep, I'll do one for Angular too. Let's do that on separate PRs though, in case there are `useAuthenticator` gotchas in other frameworks. (and tbh I'm scared I broke something this PR and have to revert it)

File: packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx
Comment (wlee221) [reply packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx:96]: Ah good question, I'll add a code comment for this as well.

Whenever you pass a specific `selector` to `useSelector`, the hook will return *only* that selected value. e.g. `useSelector(service, (state) => state.context.[...].user, comparator)` will only return `user` and not other contexts like `route` or `toResetPassword`. 

I wanted `useAuthenticator` to provide a consistent set of values, but this is up to discussions.

**Review by dbanksdesign — COMMENTED** [review]
This might be a bit out of my depth, but overall it looks good to me. Very good explanations throughout! Will this PR include updates to documentation too? I like what you have for the changeset if you want to repurpose that content.

**Comment by wlee221** [comment]
@dbanksdesign, yep documentation will be handled from #1075. I'll merge them together after both are ready/approved 👍

File: packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx
Comment (wlee221) [reply packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx]: ---
synced with @ErikCH offline. We'll adjust `useAuthenticatorUser` to just return the user. Same with other `authenticator` helpers.

File: .changeset/aero-stingrays-raise.md
Comment (wlee221) [inline .changeset/aero-stingrays-raise.md:2]: This should be minor

```suggestion
"@aws-amplify/ui-react": minor
```

File: .changeset/weak-stingrays-raise.md
Comment (wlee221) [inline .changeset/weak-stingrays-raise.md:2]: This should be minor

```suggestion
"@aws-amplify/ui-react": minor
```

**Comment by wlee221** [comment]
my css file is here as well: https://codesandbox.io/s/legacy-picker-customization-jq7wz?file=/src/App.js

**Review by ErikCH — APPROVED** [review]
Looks good!

File: packages/ui/src/helpers/auth.ts
Comment (wlee221) [inline packages/ui/src/helpers/auth.ts:220]:
```suggestion
```

File: packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx
Comment (wlee221) [inline packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx:136]:
```suggestion
```

File: packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx
Comment (wlee221) [inline packages/react/src/components/Authenticator/hooks/useAuthenticator/index.tsx:97]:
```suggestion
  /**
   * For `useSelector`'s selector argument, we just return back the `state`.
   * The reason is that whenever you select a specific value of the state, the
   * hook will return *only* that selected value instead of the whole `state`.
   *
   * To provide a consistent set of facade, we let the `selector` trivially return
   * itself and let comparator decide when to re-render.
   */
```

**Pull request merged**
